File: test/filter/go/tree.go
package main

import (
	"fmt"
	"container/list"
)

const splitThreshold = 10

// Customer represents an interval customer
type Customer struct {
	ID    int
	Start int
	End   int // Changed from Stop to End to match TypeScript
}

// TreeNode represents a tree node with linked lists
type TreeNode struct {
	Min              int
	Max              int
	Mid              int
	FullCustomers    *list.List     // Linked list for O(1) insertions
	PartialCustomers *list.List     // Linked list for O(1) insertions
	Left             *TreeNode
	Right            *TreeNode
}

// NewTreeNode creates a new tree node with pre-allocated slices
func NewTreeNode(min, max int) *TreeNode {
	return &TreeNode{
		Min:              min,
		Max:              max,
		Mid:              (min + max) / 2,
		FullCustomers:    list.New(),
		PartialCustomers: list.New(),
		Left:             nil,
		Right:            nil,
	}
}

// Insert inserts a customer into the tree node
func (node *TreeNode) Insert(customer Customer) {
	start, end := customer.Start, customer.End

	// Full containment
	if start <= node.Min && end >= node.Max {
		node.FullCustomers.PushBack(customer)
		return
	}

	// Partial overlap
	node.PartialCustomers.PushBack(customer)

	if node.PartialCustomers.Len() > splitThreshold && node.Min != node.Max {
		// Split
		if node.Left == nil && node.Right == nil {
			node.Left = NewTreeNode(node.Min, node.Mid)
			node.Right = NewTreeNode(node.Mid+1, node.Max)
		}

		// Reinsert all partial customers into children
		var next *list.Element
		for e := node.PartialCustomers.Front(); e != nil; e = next {
			next = e.Next()
			customer := e.Value.(Customer)
			node.insertIntoChildren(customer)
			node.PartialCustomers.Remove(e)
		}
	}
}

func (node *TreeNode) insertIntoChildren(customer Customer) {
	if customer.Start <= node.Mid {
		node.Left.Insert(customer)
	}
	if customer.End > node.Mid {
		node.Right.Insert(customer)
	}
}

// Dispatch finds all customers that match the given date and calls the function for each
func (node *TreeNode) Dispatch(date int, fn func(int)) {
	// Fast path: send to fullCustomers (linked list iteration)
	for e := node.FullCustomers.Front(); e != nil; e = e.Next() {
		customer := e.Value.(Customer)
		fn(customer.ID)
	}

	// Check partialCustomers (linked list iteration)
	for e := node.PartialCustomers.Front(); e != nil; e = e.Next() {
		customer := e.Value.(Customer)
		if customer.Start <= date && date <= customer.End {
			fn(customer.ID)
		}
	}

	// Traverse children
	if node.Left != nil && date <= node.Mid {
		node.Left.Dispatch(date, fn)
	} else if node.Right != nil && date > node.Mid {
		node.Right.Dispatch(date, fn)
	}
}

// PrintTree prints the tree structure for debugging
func (node *TreeNode) PrintTree(depth int) {
	indent := ""
	for i := 0; i < depth; i++ {
		indent += "  "
	}

	fmt.Printf("%sNode [%d,%d] mid:%d\n", indent, node.Min, node.Max, node.Mid)
	fmt.Printf("%s  Full: %d customers\n", indent, node.FullCustomers.Len())
	fmt.Printf("%s  Partial: %d customers\n", indent, node.PartialCustomers.Len())

	if node.Left != nil {
		fmt.Printf("%s  Left:\n", indent)
		node.Left.PrintTree(depth + 1)
	}
	if node.Right != nil {
		fmt.Printf("%s  Right:\n", indent)
		node.Right.PrintTree(depth + 1)
	}
}

// Stats returns statistics about the tree
func (node *TreeNode) Stats() (totalNodes, totalCustomers int) {
	totalNodes = 1
	totalCustomers = node.FullCustomers.Len() + node.PartialCustomers.Len()

	if node.Left != nil {
		leftNodes, leftCustomers := node.Left.Stats()
		totalNodes += leftNodes
		totalCustomers += leftCustomers
	}
	if node.Right != nil {
		rightNodes, rightCustomers := node.Right.Stats()
		totalNodes += rightNodes
		totalCustomers += rightCustomers
	}

	return totalNodes, totalCustomers
}

File: test/filter/rust/src/tree.rs
use std::collections::LinkedList;

const SPLIT_THRESHOLD: usize = 10;

#[derive(Debug, Clone, Copy)]
pub struct Customer {
    pub id: i32,
    pub start: i32,
    pub end: i32,
}

pub enum Value {
    Partial(LinkedList<Customer>),
    Parent(Box<TreeNode>, Box<TreeNode>),
}

pub struct TreeNode {
    pub min: i32,
    pub max: i32,
    pub mid: i32,
    pub full_customers: LinkedList<Customer>,
    pub value: Value,
}

impl TreeNode {
    pub fn new(min: i32, max: i32) -> Self {
        TreeNode {
            min,
            max,
            mid: (min + max) / 2,
            full_customers: LinkedList::new(),
            value: Value::Partial(LinkedList::new()),
        }
    }

    pub fn leaves(&self) -> i32 {
        match &self.value {
            Value::Parent(left, right) => left.leaves() + right.leaves(),
            Value::Partial(_) => 1
        }
    }
    pub fn insert(&mut self, customer: Customer) {
        let start = customer.start;
        let end = customer.end;

        // Full containment
        if start <= self.min && end >= self.max {
            self.full_customers.push_back(customer);
            return;
        }

        // Partial overlap
        match &mut self.value {
            Value::Parent(ref mut left, ref mut right) => {
                if start <= self.mid {
                    left.insert(customer);
                }
                if end > self.mid {
                    right.insert(customer);
                }
                return;
            }
            Value::Partial(ref mut list) => {
                list.push_back(customer);

                if list.len() > SPLIT_THRESHOLD && self.min != self.max {
                    // Split
                    let old_list = std::mem::replace(list, LinkedList::new());
                    self.value = Value::Parent(
                        Box::new(TreeNode::new(self.min, self.mid)),
                        Box::new(TreeNode::new(self.mid + 1, self.max)),
                    );

                    for customer in old_list {
                        self.insert_into_children(customer);
                    }
                }
            }
        }
    }

    fn insert_into_children(&mut self, customer: Customer) {
        if let Value::Parent(ref mut left, ref mut right) = self.value {
            if customer.start <= self.mid {
                left.insert(customer);
            }
            if customer.end > self.mid {
                right.insert(customer);
            }
        }
    }

    pub fn dispatch<F>(&self, date: i32, callback: &mut F)
    where
        F: FnMut(i32),
    {
        // Fast path: send to full_customers
        for customer in &self.full_customers {
            callback(customer.id);
        }

        // Handle based on value type
        match &self.value {
            Value::Partial(list) => {
                // Check partial_customers
                for customer in list {
                    if customer.start <= date && date <= customer.end {
                        callback(customer.id);
                    }
                }
            }
            Value::Parent(left, right) => {
                // Traverse children
                if date <= self.mid {
                    left.dispatch(date, callback);
                }
                if date > self.mid {
                    right.dispatch(date, callback);
                }
            }
        }
    }
}
